core/src/apps/webauthn/__init__.py

```diff
@@ -749,3 +1230,413 @@ def msg_error(cid: int, code: int) -> Cmd:
 
 def cmd_error(cid: int, code: int) -> Cmd:
     return Cmd(cid, _CMD_ERROR, ustruct.pack(">B", code))
+
+
+def cbor_error(cid: int, code: int) -> Cmd:
+    return Cmd(cid, _CMD_CBOR, ustruct.pack(">B", code))
+
+
+def cbor_make_credential(req: Cmd, dialog_mgr: DialogManager) -> Optional[Cmd]:
+    if not storage.is_initialized():
+        if __debug__:
+            log.warning(__name__, "not initialized")
+        return cbor_error(req.cid, _ERR_OPERATION_DENIED)
+
+    try:
+        param = cbor.decode(req.data[1:])
+        rp_id = param[_MAKECRED_CMD_RP]["id"]
+        rp_id_hash = hashlib.sha256(rp_id).digest()
+
+        # Prepare the new credential.
+        user = param[_MAKECRED_CMD_USER]
+        cred = Fido2Credential()
+        cred.rp_id = rp_id
+        cred.rp_id_hash = rp_id_hash
+        cred.rp_name = param[_MAKECRED_CMD_RP].get("name", None)
+        cred.user_id = user["id"]
+        cred.user_name = user.get("name", None)
+        cred.user_display_name = user.get("displayName", None)
+
+        # Check if any of the credential descriptors in the exclude list belong to this authenticator.
+        exclude_list = param.get(_MAKECRED_CMD_EXCLUDE_LIST, [])
+        for credential_descriptor in exclude_list:
+            excl_cred = Credential.from_bytes(credential_descriptor["id"], rp_id_hash)
+            if credential_descriptor["type"] == "public-key" and excl_cred is not None:
+                # This authenticator is already registered.
+                if not dialog_mgr.set_state(
+                    Fido2ConfirmExcluded(req.cid, dialog_mgr.iface, cred)
+                ):
+                    return cmd_error(req.cid, _ERR_CHANNEL_BUSY)
+                return None
+
+        # Check that the relying party supports ECDSA P-256 with SHA-256. We don't support any other algorithms.
+        pub_key_cred_params = param[_MAKECRED_CMD_PUB_KEY_CRED_PARAMS]
+        if _COSE_ALG_ES256 not in (alg.get("alg", None) for alg in pub_key_cred_params):
+            return cbor_error(req.cid, _ERR_UNSUPPORTED_ALGORITHM)
+
+        # Get extensions.
+        cred.hmac_secret = param.get(_MAKECRED_CMD_EXTENSIONS, {}).get(
+            "hmac-secret", False
+        )
+
+        # Get options.
+        options = param.get(_MAKECRED_CMD_OPTIONS, {})
+        resident_key = options.get("rk", False)
+        user_verification = options.get("uv", False)
+
```

Comment:
If the token is not able to perform user verification (a pin is not set) or does not support resident keys, you should return CTAP2_ERR_UNSUPPORTED_OPTION right here, that is before potentially returning CTAP2_ERR_PIN_AUTH_INVALID

Reply:
ACK